src/quadratic_polyline2_cost.cpp

```diff
@@ -76,37 +80,42 @@ void QuadraticPolyline2Cost::Quadraticize(const VectorXf& input, MatrixXf* hess,
   const Point2 current_position(input(xidx_), input(yidx_));
 
   bool is_vertex;
+  bool is_endpoint;
   LineSegment2 segment(Point2(0.0, 0.0), Point2(1.0, 1.0));
   const Point2 closest_point =
-      polyline_.ClosestPoint(current_position, &is_vertex, &segment, nullptr);
+      polyline_.ClosestPoint(current_position, &is_vertex, &segment, nullptr, &is_endpoint);
 
+  //First check whether the closest point is a endpoint of the polyline.
+  if(!is_endpoint){
   // Handle cases separately depending on whether or not closest point is
-  // a vertex of the polyline.
-  if (!is_vertex) {
-    const Point2 relative = current_position - segment.FirstPoint();
-    const Point2& unit_segment = segment.UnitDirection();
-
-    // Handle Hessian first.
-    (*hess)(xidx_, xidx_) += weight_ * unit_segment.y() * unit_segment.y();
-    (*hess)(yidx_, yidx_) += weight_ * unit_segment.x() * unit_segment.x();
-
-    const float cross_term = weight_ * unit_segment.x() * unit_segment.y();
-    (*hess)(xidx_, yidx_) -= cross_term;
-    (*hess)(yidx_, xidx_) -= cross_term;
-
-    // Handle gradient.
-    const float w_cross = weight_ * (relative.x() * unit_segment.y() -
-                                     relative.y() * unit_segment.x());
-
-    (*grad)(xidx_) += w_cross * unit_segment.y();
-    (*grad)(yidx_) -= w_cross * unit_segment.x();
-  } else {
-    (*hess)(xidx_, xidx_) += weight_;
-    (*hess)(yidx_, yidx_) += weight_;
-
-    (*grad)(xidx_) += weight_ * (current_position.x() - closest_point.x());
-    (*grad)(yidx_) += weight_ * (current_position.y() - closest_point.y());
-  }
+  // a vertex of the polyline.    
+    if (!is_vertex) {
+      const Point2 relative = current_position - segment.FirstPoint();
+      const Point2& unit_segment = segment.UnitDirection();
+
+      // Handle Hessian first.
+      (*hess)(xidx_, xidx_) += weight_ * unit_segment.y() * unit_segment.y();
+      (*hess)(yidx_, yidx_) += weight_ * unit_segment.x() * unit_segment.x();
+
+      const float cross_term = weight_ * unit_segment.x() * unit_segment.y();
+      (*hess)(xidx_, yidx_) -= cross_term;
+      (*hess)(yidx_, xidx_) -= cross_term;
+
+      // Handle gradient.
+      const float w_cross = weight_ * (relative.x() * unit_segment.y() -
+                                       relative.y() * unit_segment.x());
+
+      (*grad)(xidx_) += w_cross * unit_segment.y();
+      (*grad)(yidx_) -= w_cross * unit_segment.x();
+    } else {
+      (*hess)(xidx_, xidx_) += weight_;
+      (*hess)(yidx_, yidx_) += weight_;
+
+      (*grad)(xidx_) += weight_ * (current_position.x() - closest_point.x());
+      (*grad)(yidx_) += weight_ * (current_position.y() - closest_point.y());   
+    }
+  } else
+    return; // endpoint fix - when cost is 0
```

Comment:
just nix the `else return` since it's unnecessary

Reply:
removed.